tests/api/metrics/test_metrics_spans.py

```diff
@@ -0,0 +1,288 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import time
+import uuid
+
+import pytest
+from snowflake.cli.api.metrics import (
+    CLIMetrics,
+    CLIMetricsInvalidUsageError,
+    CLIMetricsSpan,
+)
+
+
+# helper for testing span depth limit edge case
+def create_nested_spans_recursively(metrics: CLIMetrics, num_spans: int = 1) -> None:
+    @metrics.start_span("nested_span")
+    def create_span():
+        nonlocal num_spans
+        num_spans -= 1
+
+        if num_spans > 0:
+            create_span()
+
+    create_span()
+
+
+# helper for testing span total limit edge case
+def create_spans_sequentially(metrics: CLIMetrics, num_spans: int = 1) -> None:
+    while num_spans > 0:
+        with metrics.start_span("sequential_span"):
+            pass
+        num_spans -= 1
+
+
+def test_metrics_spans_initialization_empty():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    assert metrics.current_span is None
+
+    # then
+    assert metrics.completed_spans == []
+    assert metrics.num_spans_past_total_limit == 0
+    assert metrics.num_spans_past_depth_limit == 0
+
+
+def test_metrics_spans_single_span_no_error_or_parent():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    time.sleep(0.1)
+    with metrics.start_span("span1") as span1:
+        time.sleep(0.1)
+        assert metrics.current_span is span1
+
+    assert metrics.current_span is None
+
+    # then
+    assert len(metrics.completed_spans) == 1
+    span1_dict = metrics.completed_spans[0]
+
+    assert uuid.UUID(
+        hex=span1_dict[CLIMetricsSpan.ID_KEY]
+    )  # will raise ValueError if not valid uuid
+    assert span1_dict[CLIMetricsSpan.NAME_KEY] == "span1"
+    assert span1_dict[CLIMetricsSpan.START_TIME_KEY] > 0
+    assert span1_dict[CLIMetricsSpan.EXECUTION_TIME_KEY] > 0
+    assert span1_dict[CLIMetricsSpan.ERROR_KEY] is None
+    assert span1_dict[CLIMetricsSpan.PARENT_KEY] is None
+    assert span1_dict[CLIMetricsSpan.PARENT_ID_KEY] is None
+
+
+def test_metrics_spans_finish_early_is_idempotent():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.start_span("span1") as span1:
+        start_time = span1.start_time
+        span1.finish()
+        execution_time = span1.execution_time
+
+    # then
+    assert len(metrics.completed_spans) == 1
+    span1_dict = metrics.completed_spans[0]
+    assert span1_dict[CLIMetricsSpan.START_TIME_KEY] == start_time
+    assert span1_dict[CLIMetricsSpan.EXECUTION_TIME_KEY] == execution_time
+
+
+def test_metrics_spans_parent_with_one_child():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.start_span("parent") as parent:
+        assert metrics.current_span is parent
+        # on windows, these sleeps are required to avoid the steps ending up on the same
+        # timestamps/monotonic starts and ensure a deterministic order for the completed steps
+        time.sleep(0.1)
```

Comment:
I ran into an issue with testing on windows where `time.monotonic()` only has 7 significant digits of precision, causing parent/child to be sorted in a non-deterministic and flaky order since they could have the same start time. I tried to remedy this by introducing `time.time()` as another tiebreaker for sorting, but even that ended up being the same for both spans during my tests

To remedy this, I added some sleeps in the tests which I believe simulates more realistic scenarios (since we probably won't be starting spans immediately after each other) and allows us to do a couple of things:
- Keep implementation simple
- Test whether we have the proper execution/start time inequalities (> instead of >=) between spans

In total, we have 0.7s of sleep time in these tests, which I believe is a negligible amount and shouldn't make our tests significantly longer

Let me know if you have any thoughts on alternative approaches!

Reply:
we could also use https://github.com/spulec/freezegun

Reply:
From my understanding that library freezes the values returned by the `time` functions? I'm not sure if that solves the issue that two calls to `time.monotonic()` on windows in short succession could return the same value, which breaks the sort key, but maybe I'm misunderstanding the library?

Reply:
it also allows you to fake the time by setting it to a fixed value or advancing/rewinding it at will

Reply:
for this fast of a sleep it's ok, just wanted to make you aware of the lib since it can come in handy

Reply:
Sorry, unresolving this. Relying on actual time + sleep in tests is a red flag, and almost always ends up having negative side effects down the line. We don't need to include a special library, we should be able to get away with mocking directly here. I've also used dep injection in the past for testing stuff like this, but it's probably overkill here.

Reply:
Good callout, added some [mocks for time.monotonic in the tests](https://github.com/snowflakedb/snowflake-cli/pull/1788/files#diff-839165dddafe1dd908ef843269b12e1cd4a7158ea95e5e7a2ee210ea7ec56de0R163-R177), had to convert the default_factory of the dataclass fields into lambdas instead of direct references so that they could be patched properly